iml-agent/src/action_plugins/stratagem/action_cloudsync.rs

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) 2020 DDN. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+extern crate futures;
+extern crate tokio;
+
+use crate::agent_error::{ImlAgentError, RequiredError};
+use crate::lustre::search_rootpath;
+use iml_wire_types::{FidError, FidItem};
+use liblustreapi::LlapiFid;
+use tokio::process::Command;
+use tracing::error;
+
+fn get_unique(input: &str) -> String {
+    let mut output = String::new();
+    for c in input.chars() {
+        match c {
+            '/' => output.push_str("_"),
+            _ => output.push(c),
+        }
+    }
+    output
+}
+
+async fn archive_fids(
+    llapi: LlapiFid,
+    target_name: &String,
+    fid_list: Vec<FidItem>,
+) -> Result<Vec<FidError>, ImlAgentError> {
+    let mut result = Vec::new();
+
+    for fid in fid_list {
```

Comment:
Is there a benefit to running these serially instead of concurrently?

Reply:
I hadn't figured out how to get them to run concurrently when I did this.  Filesync has concurrent code, once I get it hammered out, I'll mirror it here.

Reply:
I've updated this fn to run concurrently:

```rust
async fn archive_fids(
    llapi: LlapiFid,
    target_name: &str,
    fid_list: Vec<FidItem>,
) -> Result<Vec<FidError>, ImlAgentError> {
    stream::iter(fid_list)
        .map(|fid| {
            let fid_path = llapi.fid2path(&fid.fid);
            let mnt_pnt = llapi.mntpt();

            async move {
                let fid_path = fid_path?;
                let src_file = format!("{}/{}\n", mnt_pnt, &fid_path);

                let output = Command::new("stgm_cloudsync")
                    .arg("push")
                    .arg(src_file)
                    .arg(format!("{}:{}", target_name, get_unique(&fid_path)))
                    .output()
                    .await?;

                let FidItem { fid, data } = fid;

                Ok::<_, ImlAgentError>(FidError {
                    fid,
                    data,
                    errno: output.status.code().unwrap_or(0) as i16,
                })
            }
        })
        .chunks(10)
        .map(Ok)
        .try_fold(vec![], |mut acc, xs| async {
            let xs = try_join_all(xs).await?;

            acc.extend(xs);

            Ok::<_, ImlAgentError>(acc)
        })
        .await
}
```

I don't know how many of these will be sent in a bundle, so I used the `.chunks` combinator to limit concurrency to 10.

This fn will run at most 10 `stgm_cloudsync` at a time, all on the same task.

Reply:
Done (I believe with a modification that will handle fid2path erroring and not do the copy.